### ISSUE TITLE
MalformedPDFError on a PDF created with an Adobe product

### ISSUE THREAD

**Issue by earthlingworks** [gettalong/hexapdf#105]
Hello! We have a strange one here. Error we're getting back in the log says:

> HexaPDF::MalformedPDFError: PDF malformed around position 3388: Dictionary keys must be PDF name objects

Yes, this PDF was created by an Adobe product: https://share.getcloudapp.com/geuW1jkP

And here is an example PDF (I extracted the specific page from that was causing the issue): https://share.getcloudapp.com/5zu1bWWj

**Comment by gettalong** [comment]
Hmm.... How did you extract the example PDF? HexaPDF can load and modify it without problems...

**Comment by gettalong** [comment]
As for the error: The PDF specification only allows "name objects" (Ruby Symbol) as dictionary (Ruby Hash) keys.

What I *could* implement is the conversion of an invalid key into a symbol using something like `value.to_s.intern`. The result *might* be correct (i.e. a key that is allowed by the dictionary in question) but that is certainly not guaranteed.

Btw. if you want, you can send the malformed PDF to <info@gettalong.at> where you can be assured that the PDF is handled confidentially and deleted once analyzed.

**Comment by earthlingworks** [comment]
Gotcha. Strange, thought I tested the extracted page - sorry about that. Yeah, I can email you the malformed PDF so that we keep it confidential.

Will be able to do that later today. I appreciate the help!

Ruben
------------------------
Founder, http://www.bidsketch.com
Twitter: http://twitter.com/bidsketch

**Comment by gettalong** [comment]
Hmm... I'm still not able to reproduce. I'm running `hexapdf ins input.pdf r` and `hexapdf opt input.pdf output.pdf --compress-pages` on the file which should parse all objects.

When manually inspecting the file at position 3388 I also don't see anything out of the ordinary.

Do you have some code with which to reproduce the issue?

**Comment by earthlingworks** [comment]
Doh, sorry, I totally lost track of this thread (and juggling multiple projects at the moment).

Ok, let me take a bit of time and reproduce this with some sample code that you can try out. There must be something extra going on here.

Ruben
------------------------
Founder, http://www.bidsketch.com
Twitter: http://twitter.com/bidsketch

**Comment by earthlingworks** [comment]
Ok, I got it. I also was not able to duplicate it at first, but then I checked our original code and added the write statement we have there. I was able to duplicate it once I added "optimize: true" with that same file.

Ruben
------------------------
Founder, http://www.bidsketch.com
Twitter: http://twitter.com/bidsketch

**Comment by gettalong** [comment]
@earthlingworks Sorry, I didn't get notified of your updates, just saw it now, will have a look at this today.

**Comment by gettalong** [comment]
I tried a few things but I didn't get the error. The simplest thing I tried was:

~~~ ruby
require 'hexapdf'

HexaPDF::Document.open(ARGV[0]) do |doc|
  doc.write('gh105-out.pdf', optimize: true)
end
~~~

Which just loads and writes the file with optimization turned on but leads to no error.

Would it be possible for you to share the sample code that reproduces the error?

**Comment by earthlingworks** [comment]
Sorry about the delay. Sure thing, here's the code I used with that document to duplicate the issue:

    pdf = HexaPDF::Document.new(io: open('/path/to/pdf-original.pdf'))
    font = pdf.fonts.add('/path/to/font/OpenSans-Regular.ttf')
   
    pdf.pages.each do |page|
      overlay = page.canvas(:type => :overlay)
      box = page.box(:media)
      fragment = HexaPDF::Layout::TextFragment.create("THIS IS SOME TEST TEXT", font: font, font_size: 40)
      layouter = HexaPDF::Layout::TextLayouter.new(align: :left, font: font)
      result = layouter.fit([fragment], box.width - 300, box.height)
      result.draw(overlay, 100, 350)
    end

    pdf.write('/path/to/pdf.pdf', update_fields: true, validate: false, optimize: true)

**Comment by gettalong** [comment]
@earthlingworks Thanks!

**Comment by gettalong** [comment]
@earthlingworks Thanks for the code, I was able to reproduce the problem. Will now look for a solution.

**Comment by gettalong** [comment]
@earthlingworks I found the bug and will push a bug fix release.

**Comment by earthlingworks** [comment]
That's great to hear; thank you!

**Comment by gettalong** [comment]
@earthlingworks Just pushed a new release 0.11.7 with a fix for this problem.

i couldn't reproduce the original error message

> HexaPDF::MalformedPDFError: PDF malformed around position 3388: Dictionary keys must be PDF name objects

But running your test code from https://github.com/gettalong/hexapdf/issues/105#issuecomment-640281105 works fine now.

**Comment by earthlingworks** [comment]
Thanks, but after testing it I found that we still get that error when we import a newly written PDF. 

Here's the example code to duplicate that (took some work to figure out how to duplicate it the same way we're using it in our app): 

    pdf = HexaPDF::Document.new(io: open('/path/to/pdf-original.pdf'))
    font = pdf.fonts.add('/path/to/font/OpenSans-Regular.ttf')
   
    pdf.pages.each do |page|
      overlay = page.canvas(:type => :overlay)
      box = page.box(:media)
      fragment = HexaPDF::Layout::TextFragment.create("THIS IS SOME TEST TEXT", font: font, font_size: 40)
      layouter = HexaPDF::Layout::TextLayouter.new(align: :left, font: font)
      result = layouter.fit([fragment], box.width - 300, box.height)
      result.draw(overlay, 100, 350)
    end

    pdf.write('/path/to/pdf.pdf', update_fields: true, validate: false, optimize: true)

    target = HexaPDF::Document.new
    HexaPDF::Document.open(open('/path/to/pdf.pdff')).pages.each do |page|
      target.pages << target.import(page)
    end
    target.write('/path/to/pdf2.pdf', update_fields: true, validate: false, optimize: true)

**Comment by gettalong** [comment]
Thanks for the update! And sorry that it is still not working!

I just did a quick check to see if I'm able to reproduce and I do get now the `PDF malformed around position 478: Dictionary keys must be PDF name objects (HexaPDF::MalformedPDFError)` as in the original issue message.

Will keep you posted!

**Comment by gettalong** [comment]
@earthlingworks Found the problem. The PDF specification allows empty name (Symbol) objects, i.e. `:""` is a valid name object that is serialized to `/` (see PDF 1.7 spec section 7.3.5 last paragraph before Note 2).

HexaPDF is already aware of this and has appropriate tests in place. However, when writing out composite objects like hashes or arrays, HexaPDF performs a size optimization avoiding unnecessary spacing. And this doesn't work if `/` is a hash key or inside an array.

The PDF you provided uses empty names as hash keys. So writing out doesn't produce a segfault but it doss produce an invalid PDF.

New release is coming up...

**Comment by gettalong** [comment]
@earthlingworks Just released 0.11.8 with the fix!

**Comment by earthlingworks** [comment]
@gettalong Perfect...works great. Thanks the help on this!
